### ISSUE TITLE
Admission controller fails on timeout when failurePolicy set to Ignore

### ISSUE THREAD

**Issue by omri86** [kubernetes/kubernetes#71508]
Trying to set up a validating admission webhook on my GKE cluster using the following yaml:

```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingWebhookConfiguration
metadata:
  name: test-admission-webhook
webhooks:
  - name: my-admission-webhook.io
    rules:
      - apiGroups:
          - ""
        apiVersions:
          - "v1"
        operations:
          - "CREATE"
        resources:
          - "pods"
    failurePolicy: Ignore
    clientConfig:
      url: "https://192.168.99.1:8080"
      caBundle: %%TEST_BUNDLE%%
```

And my server is still down, I can't run any pod on my cluster, getting the following error:

```
Error creating: Timeout: request did not complete within allowed duration
```

Even when I removed the `failurePolicy` field from my yaml file (which is supposed to default to `Ignore` as mentioned in the [official doc](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#webhook-v1beta1-admissionregistration-k8s-io)) and I'm getting the same error.

/sig api-machinery
/kind bug

**Comment by yue9944882** [comment]
> Error creating: Timeout: request did not complete within allowed duration

this will happen when calling webhook exceeds request timeout and the cause is that ur webhook is not responding.

**Comment by omri86** [comment]
@yue9944882 Thanks for your reply.

I know that this happens because my webhook is not responding *but* from what I understand the `failurePolicy` field defines what happens in this kind of result:
```
FailurePolicy defines how unrecognized errors from the admission endpoint are handled - allowed values are Ignore or Fail. Defaults to Ignore.
```

So when setting it to `Ignore` - the pod should be up and running even though the webhook is not responding, am I wrong?

**Comment by yue9944882** [comment]
> So when setting it to Ignore - the pod should be up and running even though the webhook is not responding, am I wrong?

no, actually, it's ur CREATE request failing not ur admission request, which is to say, ur CREATE request failed before apiserver getting know and ignore that response from ur webhook.  you can workaround by setting a shorter timeout in ur webhook i suppose

**Comment by yue9944882** [comment]
@kubernetes/sig-api-machinery-feature-requests do you think about adding a timeout option to webhook configuration?

/kind feature

**Comment by omri86** [comment]
@yue9944882  Thanks again for your help - but still I'm not sure I follow.

I've set up the webhook to some address, then I'm trying to send the CREATE request to that should arrive to that webhook - what you're saying is that as long as the webhook is down, every CREATE request sent will fail? 
This seems a bit odd - is there any way around this? how will reducing the timeout helps in this scenario?

And BTW the feature for timeout on webhook configuration already exist here: #60914

**Comment by yue9944882** [comment]
> So when setting it to Ignore - the pod should be up and running even though the webhook is not responding, am I wrong?

trying to explain why your requests failed. the `failurePolicy` defines how we deal w/ responses returned from webhooks, while apiserver doesn't receive anything if the webhook is not responding so the `failurePolicy` didn't work. currently calling webhooks doesn't have a timeout but we have a __generic timeout__ (defaulted to 60s) for all in-coming requests which is the cause why you can't run any new pods in your case. does that make sense? 

> is there any way around this? how will reducing the timeout help in this scenario?

to clarify, i suggest setting a timeout in your webhook server too so that it could return error status like 5XX instead of un-responding. To walkaround that, do you think about changing the matching rules to bypass pod creation? Or just simply delete that webhook before it's set. Would that help in your scenario?

**Comment by liggitt** [comment]
> the failurePolicy defines how we deal w/ responses returned from webhooks, while apiserver doesn't receive anything if the webhook is not responding so the failurePolicy didn't work

It should work. A failure policy of ignore should fail open on timeout or other call errors

**Comment by yue9944882** [comment]
then we need a proper timeout here

[https://github.com/kubernetes/kubernetes/blob/c2d93cb1b231b5d2ec58d722fc6d34719a88e3a7/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go#L167-L168](https://github.com/kubernetes/kubernetes/blob/c2d93cb1b231b5d2ec58d722fc6d34719a88e3a7/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go#L167-L168)

**Comment by gyuho** [comment]
Any updates?

**Comment by roycaihw** [comment]
it's not necessary that admission webhook is causing the failure

- a failure policy of ignore should fail open on timeout or other call errors. We had tested [ignoring webhook that cannot talk to server works](https://github.com/kubernetes/kubernetes/blob/98fa2c7d32e59f798ff4201dfdcb439554392beb/test/e2e/apimachinery/webhook.go#L394-L397)
- [configurable timeout for webhook](https://github.com/kubernetes/kubernetes/pull/74562) is added in 1.14. We also tested that [a ignore failure policy works when a configured timeout occurs](https://github.com/kubernetes/kubernetes/blob/98fa2c7d32e59f798ff4201dfdcb439554392beb/test/e2e/apimachinery/webhook.go#L209-L217)
- I see the error could also be caused by initializer: https://github.com/kubernetes/kubernetes/issues/54228, https://github.com/kubernetes/kubernetes/issues/51485, https://github.com/Azure/acs-engine/issues/3187

@omri86 I suspect if you see the error because of components other than admission webhook, but it's hard to tell with the given information

**Comment by omri86** [comment]
@roycaihw Do you need me to supply more data?

**Comment by floriankoch** [comment]
@roycaihw  we also hit this bug, i our case , we run the admission controller on kubernetes, and when updating a node ( on wich the admission controller runs) , the networking (calico) does not come up , because the admission controller was down, and this bug happen , so the failure policy does not work

**Comment by liggitt** [comment]
> the failure policy does not work

how long does the API request take before failing? it is possible you are encountering https://github.com/kubernetes/kubernetes/issues/60914#issuecomment-455149732, where the client-side timeout aborts the request before the webhook timeout is reached

**Comment by roycaihw** [comment]
I think @yue9944882 and @liggitt were right. It's your create request timing-out not your admission request

tl;dr: it's behaving by design. But we shouldn't use the same timeout for client request and admission request. You could do one of the following to fix: 

- configure timeout (> 30s) in your client request by setting `Timeout` in your `restclient.Config` (e.g. `config` [here](https://github.com/kubernetes/kubernetes/blob/9eebfe7a04e59668adf31bafbaa38db56527dda8/staging/src/k8s.io/client-go/examples/create-update-delete-deployment/main.go#L58)). It will change the timeout for all your client requests (per request timeout configuration is [WIP](https://github.com/kubernetes/kubernetes/pull/53076)) 
- configure timeout (< 30s) in your webhook server as @yue9944882 suggested
- configure timeout (< 30s) for admission request using https://github.com/kubernetes/kubernetes/pull/74562 (it's in 1.14, probably the least solution you want)

(longer version) I think what happened is:

there are two different requests, built on same client package with the same timeout
1. you client sends a create request to apiserver using client-go (which builds on [rest client](https://github.com/kubernetes/kubernetes/blob/47063891dd782835170f500a83f37cc98c3c1013/staging/src/k8s.io/client-go/rest/client.go#L61), and eventually on a [http client](https://github.com/kubernetes/kubernetes/blob/47063891dd782835170f500a83f37cc98c3c1013/staging/src/k8s.io/client-go/rest/client.go#L80)). The http client has a [timeout set for every request](https://github.com/kubernetes/kubernetes/blob/47063891dd782835170f500a83f37cc98c3c1013/staging/src/k8s.io/client-go/rest/config.go#L294-L302) (I think the config is defaulted to 30s somewhere)
2. apiserver receives the create request from client and sends an admission request to webhook server, also using [rest client](https://github.com/kubernetes/kubernetes/blob/47063891dd782835170f500a83f37cc98c3c1013/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go#L118) with 30s timeout

since webhook server is unresponsive, both requests hang. 

3. your client hits its timeout and returns error first
4. apiserver hits timeout talking to webhook server. It could have ignored the error based on the policy and created pod successfully, but the client has dropped alreay 

(you can tell from the error message, it should contain the text ["Internal error"](https://github.com/kubernetes/kubernetes/blob/release-1.11/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go#L71-L72) if apiserver actually didn't honor ignore policy and returned error)

**Comment by roycaihw** [comment]
/close

as it's working as expected
